core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala

```diff
@@ -148,14 +148,17 @@ class HistoryServer(
       appId: String,
       attemptId: Option[String],
       ui: SparkUI,
-      completed: Boolean) {
+      completed: Boolean): Unit = this.synchronized {
```

Comment:
Hmm... would it be better to synchronize `SparkUI.attachHandler` and `SparkUI.detachHandler`? Or maybe restrict the scope of the synchronization here?

This seems too coarse; in the detach case, it synchronizes around the `onUIDetached` callback which can do expensive locking and I/O.

Reply:
In order to fine grain the synchronization, can we add synchronize on handlers only?